File: frontend/src/lib/types/agent-runs.ts
import type {
  AgentQueryResponse,
  AgentStepResponse,
  CitationResponse,
} from "@/lib/types/agent";

export type AgentRunStatus = "success" | "failure";

export interface AgentRunSummary {
  run_id: string;
  query: string;
  status: AgentRunStatus;
  started_at: string;
  completed_at: string;
  duration_ms: number | null;
  tool_used: string | null;
  step_count: number;
  citation_count: number;
  error_message: string | null;
  error_code: string | null;
}

export interface AgentRunListResponse {
  runs: AgentRunSummary[];
  total: number;
  limit: number;
  offset: number;
}

export interface AgentRunDetail extends AgentRunSummary {
  answer: string | null;
  citations: CitationResponse[];
  steps: AgentStepResponse[];
  metadata: Record<string, unknown>;
}

export interface AgentRunListParams {
  search?: string;
  status?: AgentRunStatus;
  limit?: number;
  offset?: number;
}

/** Shape compatible with AgentQueryResponse for reusing trace components. */
export type AgentRunTraceData = Pick<
  AgentQueryResponse,
  "answer" | "citations" | "tool_used" | "steps" | "metadata"
>;

export function agentRunDetailToTraceData(run: AgentRunDetail): AgentRunTraceData {
  return {
    answer: run.answer ?? "",
    citations: run.citations,
    tool_used: run.tool_used,
    steps: run.steps,
    metadata: run.metadata,
  };
}

File: app/schemas/agent_runs.py
"""Agent run history API schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.agent import AgentStepResponse
from app.schemas.query import CitationResponse


class AgentRunSummaryResponse(BaseModel):
    """Summary item for agent run listing."""

    run_id: str
    query: str
    status: Literal["success", "failure"]
    started_at: datetime
    completed_at: datetime
    duration_ms: int | None = None
    tool_used: str | None = None
    step_count: int = 0
    citation_count: int = 0
    error_message: str | None = None
    error_code: str | None = None


class AgentRunListResponse(BaseModel):
    """Paginated agent run history."""

    runs: list[AgentRunSummaryResponse]
    total: int
    limit: int
    offset: int


class AgentRunDetailResponse(AgentRunSummaryResponse):
    """Full stored agent run payload."""

    answer: str | None = None
    citations: list[CitationResponse] = Field(default_factory=list)
    steps: list[AgentStepResponse] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
